File: Java/src/gov/nasa/larcfm/ACCoRD/CDCylinder.java
/*
 * CD3D.java 
 * Release: ACCoRDj-2.b (08/22/10) 
 *
 * Contact: Cesar Munoz (cesar.a.munoz@nasa.gov)
 * NASA LaRC
 * http://shemesh.larc.nasa.gov/people/cam/ACCoRD
 *
 * CD3D is an algorithm for 3-D conflict *detection*.
 *
 * Unit Convention
 * ---------------
 * All units in this file are *internal*:
 * - Units of distance are denoted [d]
 * - Units of time are denoted     [t]
 * - Units of speed are denoted    [d/t]
 *
 * REMARK: X Vect3s to East, Y Vect3s to North. 
 *
 * Naming Convention
 * -----------------
 *   The intruder is fixed at the origin of the coordinate system.
 * 
 *   D  : Diameter of the protected zone [d]
 *   H  : Height of the protected zone [d]
 *   B  : Lower bound of lookahed time interval [t] (B >= 0)
 *   T  : Upper bound of lookahead time interval [t] (T < 0 means infinite lookahead time)
 *   s  : Relative 3-D position of the ownship [d,d,d]
 *   vo : Ownship velocity vector [d/t,d/t,d/t]
 *   vi : Traffic velocity vector [d/t,d/t,d/t]
 * 
 * Functions
 * ---------
 * violation : Check for 3-D loss of separation
 * detection : 3-D conflict detection with calculation of conflict interval 
 * cd3d      : Check for predicted conflict
 * 
 * Global variables (modified by detection)
 * ----------------
 * t_in  : Time to loss of separation
 * t_out : Time to recovery of loss of separation
 * 
 * Copyright (c) 2011-2017 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 * 
 * 
 * Note: The B and T parameters also affect the time in and time out of loss values.
 * 
 */

package gov.nasa.larcfm.ACCoRD;

import gov.nasa.larcfm.Util.LossData;
import gov.nasa.larcfm.Util.ParameterData;
import gov.nasa.larcfm.Util.Vect3;
import gov.nasa.larcfm.Util.Velocity;

public class CDCylinder implements Detection3D {

  private CD3DTable table;

  private String id = "";

  /**
   * Instantiates a new CD3D object.
   */
  public CDCylinder() {
    table = new CD3DTable();
  }

  /**
   * This specifies the internal table is a copy of the provided table
   * @param tab
   */
  public CDCylinder(CD3DTable tab) {
    table = tab.copy();
  }

  public CDCylinder(double d, String dunit, double h, String hunit) {
    table = new CD3DTable(d,dunit,h,hunit);
  }

  /**
   * Create a new state-based conflict detection object using specified units.
   * 
   * @param distance the minimum horizontal separation distance in specified units
   * @param height the minimum vertical separation height in specified units.
   */
  public static CDCylinder make(double distance, String dUnits, double height, String hUnits) {
    return new CDCylinder(distance,dUnits,height,hUnits);  
  }

  /**
   * Create a new state-based conflict detection object using internal units.
   * 
   * @param distance the minimum horizontal separation distance [m]
   * @param height the minimum vertical separation height [m].
   */
  public static CDCylinder mk(double distance, double height) {
    return new CDCylinder(distance, "m", height, "m");  
  }

  /**
   * Return a copy of this object's table
   */
  public CD3DTable getCD3DTable() {
    return table.copy();
  }

  /** Sets the internal table to be a copy of the supplied one */
  public void setCD3DTable(CD3DTable tab) {
    table = tab.copy();
  }

  public double getHorizontalSeparation() {
    return table.getHorizontalSeparation();
  }

  public double getVerticalSeparation() {
    return table.getVerticalSeparation();
  }

  public void setHorizontalSeparation(double d) {
    table.setHorizontalSeparation(d);
  }

  public void setVerticalSeparation(double h) {
    table.setVerticalSeparation(h);
  }

  public double getHorizontalSeparation(String unit) {
    return table.getHorizontalSeparation(unit);
  }

  public double getVerticalSeparation(String unit) {
    return table.getVerticalSeparation(unit);
  }

  public void setHorizontalSeparation(double d, String unit) {
    table.setHorizontalSeparation(d, unit);
  }

  public void setVerticalSeparation(double h, String unit) {
    table.setVerticalSeparation(h, unit);
  }

  /**
   * Computes the conflict time interval in [B,T].
   * 
   * @param s the relative position of the aircraft
   * @param vo the ownship's velocity
   * @param vi the intruder's velocity
   * @param D the minimum horizontal distance
   * @param H the minimum vertical distance
   * @param B the the lower bound of the lookahead time (B >= 0)
   * @param T the upper bound of the lookahead time (B < T)
   * 
   * @return true, if the conflict time interval (t_in,t_out) is in [B,T].
   */
  public LossData detection(Vect3 s, Vect3 vo, Vect3 vi, double D, double H, double B, double T) { 
    return CD3D.detection(s,vo,vi,D,H,B,T);
  }

  /**
   * Computes the conflict time interval in [0,T].
   * 
   * @param s the relative position of the aircraft
   * @param vo the ownship's velocity
   * @param vi the intruder's velocity
   * @param D the minimum horizontal distance
   * @param H the minimum vertical distance
   * @param T the the lookahead time (T > 0)
   * 
   * @return true, if the conflict time interval (t_in,t_out) is in [0,T].
   */
  public LossData detection(Vect3 s, Vect3 vo, Vect3 vi, double D, double H, double T) {
    return detection(s,vo,vi,D,H,0,T);
  }

  /**
   * Computes the conflict time interval in [0,...).
   * 
   * @param s the relative position of the aircraft
   * @param vo the ownship's velocity
   * @param vi the intruder's velocity
   * @param D the minimum horizontal distance
   * @param H the minimum vertical distance
   * 
   * @return true, if the conflict time interval (t_in,t_out) is in [0,...)
   */
  public LossData detection(Vect3 s, Vect3 vo, Vect3 vi, double D, double H) {
    return detection(s,vo,vi,D,H,0,Double.POSITIVE_INFINITY);
  }

  public String toString() {
    return (id.equals("") ? "" : id+" : ")+getSimpleClassName()+" = {"+table.toString()+"}";
  }
  
  public String toPVS(int prec) {
    return table.toPVS(prec);
  }

  public boolean violation(Vect3 so, Velocity vo, Vect3 si, Velocity vi, double D, double H) {
    return CD3D.lossOfSep(so,si,D,H);
  }


  public boolean violation(Vect3 so, Velocity vo, Vect3 si, Velocity vi) {
    return violation(so,vo,si,vi,table.getHorizontalSeparation(),table.getVerticalSeparation());
  }

  public boolean conflict(Vect3 so, Velocity vo, Vect3 si, Velocity vi, double D, double H, double B, double T) {
    return CD3D.cd3d(so.Sub(si), vo, vi, D, H, B, T); 
  }

  public boolean conflict(Vect3 so, Velocity vo, Vect3 si, Velocity vi, double B, double T) {
    return conflict(so, vo, si, vi, table.getHorizontalSeparation(), table.getVerticalSeparation(), B, T); 
  }

  public ConflictData conflictDetection(Vect3 so, Velocity vo, Vect3 si, Velocity vi, double D, double H, double B, double T) {
    Vect3 s = so.Sub(si);
    Velocity v = vo.Sub(vi);
    double t_tca = CD3D.tccpa(s, vo, vi, D, H, B, T);
    double dist_tca = s.linear(v,t_tca).cyl_norm(D, H);
    LossData ld = CD3D.detection(s,vo,vi,D,H,B,T);
    return new ConflictData(ld,t_tca,dist_tca,s,v);
  }

  public ConflictData conflictDetection(Vect3 so, Velocity vo, Vect3 si, Velocity vi, double B, double T) {
    return conflictDetection(so,vo,si,vi,table.getHorizontalSeparation(), table.getVerticalSeparation(), B, T); 
  }

  public double timeOfClosestApproach(Vect3 so, Velocity vo, Vect3 si, Velocity vi, double D, double H, double B, double T) {
    return CD3D.tccpa(so.Sub(si), vo, vi, D, H, B, T);
  }

  public double timeOfClosestApproach(Vect3 so, Velocity vo, Vect3 si, Velocity vi, double B, double T) {
    return timeOfClosestApproach(so,vo,si,vi,table.getHorizontalSeparation(), table.getVerticalSeparation(), B, T); 
  }

  public ParameterData getParameters() {
    ParameterData p = new ParameterData();
    updateParameterData(p);
    return p;
  }

  public void updateParameterData(ParameterData p) {
    table.updateParameterData(p);
    p.set("id",id);
  }

  public void setParameters(ParameterData p) {
    table.setParameters(p);
    if (p.contains("id")) {
      id = p.getString("id");
    }
  }

  /**
   * Returns a fresh instance of this type of Detection3D with default parameter data.
   */
  public CDCylinder make() {
    return new CDCylinder();
  }

  /**
   * Returns a deep copy of this CDCylinder object, including any results that have been calculated.  This will duplicate parameter data, but will NOT
   * link any existing CD3DTable.  Call setCD3DTable() if linking is necessary.
   */
  public CDCylinder copy() {
    CDCylinder ret = new CDCylinder();
    ret.id = id;
    ret.table = new CD3DTable(table);
    return ret;
  }

  public String getSimpleClassName() {
    return getClass().getSimpleName();
  }
  
  public String getCanonicalClassName() {
    return getClass().getCanonicalName(); 
  }

  public String getIdentifier() {
    return id;
  }

  public void setIdentifier(String s) {
    id = s;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    CDCylinder other = (CDCylinder) obj;
    if (id == null) {
      if (other.id != null)
        return false;
    } else if (!id.equals(other.id))
      return false;
    if (table == null) {
      if (other.table != null)
        return false;
    } else if (!table.equals(other.table))
      return false;
    return true;
  }

  public boolean contains(Detection3D cd) {
    if (cd instanceof CDCylinder) {
      CDCylinder d = (CDCylinder) cd;
      return table.D >= d.table.D && table.H >= d.table.H; 
    }
    return false;
  }

}

File: Modules/ACCoRD/src/CDCylinder.cpp
/*
 * CD3D.h 
 * Release: ACCoRD++-2.b (08/22/10) 
 *
 * Contact: Cesar Munoz (cesar.a.munoz@nasa.gov)
 * NASA LaRC
 * http://shemesh.larc.nasa.gov/people/cam/ACCoRD
 *
 * Copyright (c) 2011-2020 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

#include "CDCylinder.h"
#include "LossData.h"
#include "LossData.h"
#include "ConflictData.h"
#include "DaidalusParameters.h"
#include <cmath>

#include "CD3D.h"
#include "format.h"

namespace larcfm {

/**
 * Instantiates a new CD3D object.
 */
CDCylinder::CDCylinder(const std::string& s) {
  D_ = Units::from("nmi", 5.0);
  H_ = Units::from("ft", 1000.0);
  units_["D"] = "nmi";
  units_["H"] = "ft";
  id = s;
}

CDCylinder::CDCylinder(const CDCylinder& cdc) {
  D_ = cdc.D_;
  H_ = cdc.H_;
  units_ = cdc.units_;
  id = cdc.id;
}

CDCylinder::CDCylinder(double d, double h) {
  D_ = std::abs(d);
  H_ = std::abs(h);
  units_["D"] = "m";
  units_["H"] = "m";
  id = "";
}

CDCylinder::CDCylinder(double d, const std::string& dunit, double h, const std::string& hunit) {
  D_ = Units::from(dunit,std::abs(d));
  H_ = Units::from(hunit,std::abs(h));
  units_["D"] = dunit;
  units_["H"] = hunit;
  id = "";
}

CDCylinder CDCylinder::make(double distance, const std::string& dUnits, double height, const std::string& hUnits) {
  return CDCylinder(distance,dUnits,height,hUnits);
}

CDCylinder CDCylinder::mk(double distance, double height) {
  return CDCylinder(distance,height);
}

/**
 * @return one static CDCylinder
 */
const CDCylinder& CDCylinder::A_CDCylinder() {
  static CDCylinder cd3d;
  return cd3d;
}

/**
 * @return CDCylinder thresholds, i.e., D=5nmi, H=1000ft.
 */
const CDCylinder& CDCylinder::CD3DCylinder() {
  return A_CDCylinder();
}

std::string CDCylinder::getUnits(const std::string& key) const {
  std::map<std::string,std::string>::const_iterator got = units_.find(key);
  if (got == units_.end()) {
    return "unspecified";
  }
  return got->second;
}

double CDCylinder::getHorizontalSeparation() const {
  return D_;
}

void CDCylinder::setHorizontalSeparation(double d) {
  D_ = std::abs(d);
}

double CDCylinder::getVerticalSeparation() const {
  return H_;
}

void CDCylinder::setVerticalSeparation(double h) {
  H_ = std::abs(h);
}

double CDCylinder::getHorizontalSeparation(const std::string& u) const {
  return Units::to(u, D_);
}

void CDCylinder::setHorizontalSeparation(double d, const std::string& u) {
  setHorizontalSeparation(Units::from(u,d));
  units_["D"]=u;
}

double CDCylinder::getVerticalSeparation(const std::string& u) const {
  return Units::to(u, H_);
}

void CDCylinder::setVerticalSeparation(double h, const std::string& u) {
  setVerticalSeparation(Units::from(u,h));
  units_["H"]=u;
}

LossData CDCylinder::detection(const Vect3& s, const Vect3& vo, const Vect3& vi, const double D, const double H, const double B, const double T) const {
  return CD3D::detection(s,vo,vi,D,H,B,T);
}

LossData CDCylinder::detection(const Vect3& s, const Vect3& vo, const Vect3& vi, const double D, const double H, const double T) const {
  return detection(s,vo,vi,D,H,0.0,T);
}

LossData CDCylinder::detection(const Vect3& s, const Vect3& vo, const Vect3& vi, const double D, const double H) const {
  return detection(s,vo,vi,D,H,0.0,PINFINITY);
}

ConflictData CDCylinder::conflict_detection(const Vect3& so, const Velocity& vo, const Vect3& si, const Velocity& vi, double D, double H, double B, double T) {
  Vect3 s = so.Sub(si);
  Vect3 v = vo.Sub(vi);
  double t_tca = CD3D::tccpa(s, vo, vi, D, H, B, T);
  double dist_tca = s.linear(v,t_tca).cyl_norm(D, H);
  LossData ld = CD3D::detection(s,vo,vi,D,H,B,T);
  return ConflictData(ld,t_tca,dist_tca,s,v);
}

ConflictData CDCylinder::conflictDetection(const Vect3& so, const Velocity& vo, const Vect3& si, const Velocity& vi, double B, double T) const {
  return conflict_detection(so,vo,si,vi,D_,H_,B,T);
}

double CDCylinder::time_of_closest_approach(const Vect3& so, const Velocity& vo, const Vect3& si, const Velocity& vi, double D, double H, double B, double T) {
  return CD3D::tccpa(so.Sub(si),vo,vi,D,H,B,T);
}

double CDCylinder::timeOfClosestApproach(const Vect3& so, const Velocity& vo, const Vect3& si, const Velocity& vi, double B, double T) const {
  return time_of_closest_approach(so,vo,si,vi,D_,H_,B,T);
}

CDCylinder* CDCylinder::make() const {
  return new CDCylinder();
}

CDCylinder* CDCylinder::copy() const {
  CDCylinder* cd = new CDCylinder(*this);
  return cd;
}

ParameterData CDCylinder::getParameters() const {
  ParameterData p;
  updateParameterData(p);
  return p;
}

void CDCylinder::updateParameterData(ParameterData& p) const {
  p.setInternal("D",D_,getUnits("D"));
  p.setInternal("H",H_,getUnits("H"));
  p.set("id",id);
}

void CDCylinder::setParameters(const ParameterData& p) {
  if (p.contains("D")) {
    setHorizontalSeparation(p.getValue("D"));
    units_["D"]=p.getUnit("D");
  }
  if (p.contains("H")) {
    setVerticalSeparation(p.getValue("H"));
    units_["H"]=p.getUnit("H");
  }
  if (p.contains("id")) {
    id = p.getString("id");
  }
}

std::string CDCylinder::getSimpleClassName() const {
  return "CDCylinder";
}

std::string CDCylinder::toString() const {
  return (id == "" ? "" : id+" : ")+getSimpleClassName()+" = {D = "+
      Units::str(getUnits("D"),D_)+", H = "+Units::str(getUnits("H"),H_)+"}";
}

std::string CDCylinder::toPVS() const {
  return getSimpleClassName()+"((# D:= "+FmPrecision(D_)+", H:= "+FmPrecision(H_)+" #))";
}

std::string CDCylinder::getIdentifier() const {
  return id;
}

void CDCylinder::setIdentifier(const std::string& s) {
  id = s;
}

bool CDCylinder::equals(Detection3D* det) const {
  if (!larcfm::equals(getCanonicalClassName(), det->getCanonicalClassName())) return false;
  if (!larcfm::equals(id,det->getIdentifier())) return false;
  CDCylinder* cd = (CDCylinder*)det;
  if (D_ != cd->D_) return false;
  if (H_ != cd->H_) return false;
  return true;
}

bool CDCylinder::contains(const Detection3D* det) const {
  if (larcfm::equals(getCanonicalClassName(), det->getCanonicalClassName())) {
    CDCylinder* cd = (CDCylinder*)det;
    return D_ >= cd->D_ && H_ >= cd->H_;
  }
  return false;
}

/* Return a list of point representing a counter-clockwise circular arc centered at pc,
 * whose first point is pc+v(0), and the last one is pc+v(alpha), where alpha is
 * in [0,2*pi].
 */
void CDCylinder::circular_arc(std::vector<Position>& arc, const Position& pc, const Velocity& v,
    double alpha, bool include_last) {
  alpha = Util::almost_equals(alpha,2*Pi,DaidalusParameters::ALMOST_) ? alpha :  Util::to_2pi(alpha);
  double step = Pi/180;
  arc.push_back(pc.linear(v,1));
  double current_trk = v.trk();
  for (double a = step; Util::almost_less(a,alpha,DaidalusParameters::ALMOST_); a += step) {
    arc.push_back(pc.linear(v.mkTrk(current_trk-a),1));
  }
  if (include_last) {
    arc.push_back(pc.linear(v.mkTrk(current_trk-alpha),1));
  }
}

void CDCylinder::horizontalHazardZone(std::vector<Position>& haz,
    const TrafficState& ownship, const TrafficState& intruder, double T) const {
  haz.clear();
  const Position& po = ownship.getPosition();
  Velocity v = Velocity::make(ownship.getVelocity().Sub(intruder.getVelocity()));
  if (Util::almost_equals(T,0) || Util::almost_equals(v.norm2D(),0)) {
    circular_arc(haz,po,Velocity::mkVxyz(D_,0,0),2*Pi,false);
  } else {
    Vect3 sD = Horizontal::unit_perpL(v).Scal(D_);
    Velocity vD = Velocity::make(sD);
    Velocity vnD = Velocity::make(sD.Neg());
    circular_arc(haz,po,vD,Pi,true);
    circular_arc(haz,po.linear(v,T),vnD,Pi,true);
  }
}

}

